pyHalo/realization_extensions.py

```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
+            lens_model_at_plane = LensModel(lens_model_list_at_plane)
+            xcoords, ycoords = xx_base + center_x, yy_base + center_y
+            projected_mass = lens_model_at_plane.kappa(xcoords.ravel(), ycoords.ravel(), kwargs_lens_at_plane)
+            probability_density = projected_mass.reshape(shape0)
+            coord_x_clumpy, coord_y_clumpy = self._sample_projmass(probability_density, Nclumpy, pixel_scale, 
+                                                   center_x, center_y, max_rendering_range)
+            masses_clumpy = np.array([M] * len(coord_x_clumpy))
+            z_shifts_clumpy = [zi] * len(masses_clumpy)
+            return coord_x_clumpy, coord_y_clumpy, z_shifts_clumpy, masses_clumpy
+    
+    def _pbh_in_lensplane (self, lens_plane_redshifts, cosmology, cosmo_geometry, max_rendering_range_base, 
+                       x_interp_list, y_interp_list, image_index, LMF, mlow, mhigh, rho_DM, npix, M, realization):
+        # intializing storage arrays
+        pbh_masses = np.array([])
+        pbh_x_coordinates = np.array([])
+        pbh_y_coordinates = np.array([])
+        pbh_redshifts = []
+        
+            
+        for zi in lens_plane_redshifts:
+            
+            d_comoving = cosmology.D_C_z(zi)
+            
+            # this scales the rendering area with redshift and generates a grid
+            max_rendering_range = max_rendering_range_base * cosmo_geometry.rendering_scale(zi) 
+            
+            
+            # compute the angular coordinate around which to render halos
+            center_x, center_y = x_interp_list[image_index](d_comoving), y_interp_list[image_index](d_comoving)
+            
+            # get the realization at redshift z
+            realization_at_redshift, indexes = realization_at_z(realization, zi, center_x, center_y, max_rendering_range)
+            
+            # get an instance of LensModel at redshift z
+            args = realization_at_redshift.lensing_quantities(add_mass_sheet_correction=False)
+            lens_model_list_at_plane, kwargs_lens_at_plane = args[0], args[2]
+            
+            delta_zi = 0.02
+            #delta_zi= lens_plane_redshifts[1]-lens_plane_redshifts[0]   
+            V = cosmo_geometry.volume_element_comoving(zi, delta_z=delta_zi, radius=max_rendering_range)
+          
+            f_halos =  LMF.mass_fraction_in_halos(zi, mlow, mhigh) # this is where mlow and mhigh are necessary
+            
+            # M is black hole mass (solar masses)
+            
+            _nsmooth = f_halos*rho_DM*(V/M)
+            _nclumpy = (1-f_halos)*rho_DM*(V/M)
+            
+            Nclumpy = int(np.random.poisson(_nclumpy))
+            Nsmooth = int(np.random.poisson(_nsmooth))
+            
+    
+            if Nsmooth > 0:
+                coord_x_smooth, coord_y_smooth,  z_shifts_smooth, masses_smooth = self._smooth_pbh (max_rendering_range, Nsmooth, center_x, center_y, zi, M)
```

Comment:
regarding the comments above, this line should return the coord_x/y_smooth arrays, and the z_shifts_smooth can be defined in the line immediately following. For now, you can just create a list of masses, but if the future we might want to sample from a mass function (which is why you don't want to put that functionality where the halo coordinates are generated)